### ISSUE TITLE
CZ Quertz keyboard layout. Can I do it somehow?

### ISSUE THREAD

**Issue by orangeb1n** [CosmosOS/Cosmos#2578]
#### Area of Cosmos - What area of Cosmos are we dealing with?
Hardware communication

#### Expected Behaviour - What do you think that should happen?
I excepted more keyboard layouts (CZ)

#### Actual Behaviour - What unexpectedly happens?
4 KEYBOARD LAYOUTS. EN, FR, DE AND TR
(really... WHAT'S TR AT ALL?!)

#### Reproduction - How did you get this error to appear?
When I look up the code

#### Version - Were you using the User Kit or Dev Kit? And what User Kit version or Dev Kit commit (Cosmos, IL2CPU, X#)?
Cosmos latest, IL2CPU sorry I don't know what's that, .NET 6.0

**Comment by ascpixi** [comment]
Yes, you're welcome to contribute to the project - we'd appreciate a CZ keyboard layout! You can see an example of such a layout [here](https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.System2/Keyboard/ScanMaps/US_Standard.cs#L9) - this one is for the standard US one.

**Comment by slavekrouta** [comment]
> Currently, the type of layout you're looking for isn't available in Cosmos. You can try implementing one yourself - if you do decide to do that, consider making a pull request that adds this functionality. We would love for you to join us and contribute to the project! You can see an example of such a layout [here](https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.System2/Keyboard/ScanMaps/US_Standard.cs#L9) - this class definition is for the standard US one.

I'll take a look, but can you tell me, what are these? (for ex. in Q key's code):  _keys.Add(new KeyMapping(0x02, '1', '!', '1', '1', '!', '1', ConsoleKeyEx.D1));
tell me the elements of this line. thanks. (i know my english is bad)

**Comment by slavekrouta** [comment]
I figured it out i guess....
so (on the E key for ex.):
_keys.Add(new KeyMapping(0x12, 'e', 'E', 'e', 'E', 'e', 'E', ConsoleKeyEx.E));
the first e means key, that will appear on normal press.
the first uppercase E means key when the shift is held
the second e appears with shift held and capslock on
the second E appears with shift held and capslock on (i found this one on internet but it does not make any sense for me because shift + caps + e = e  tho....)
the third e appears with caps on (from internet again, no sense again)
the third E appears with caps on
Prove me wrong thanks and explain the second E and third e

**Comment by ascpixi** [comment]
> > Currently, the type of layout you're looking for isn't available in Cosmos. You can try implementing one yourself - if you do decide to do that, consider making a pull request that adds this functionality. We would love for you to join us and contribute to the project! You can see an example of such a layout [here](https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.System2/Keyboard/ScanMaps/US_Standard.cs#L9) - this class definition is for the standard US one.
> 
> I'll take a look, but can you tell me, what are these? : _keys.Add(new KeyMapping(0x02, '1', '!', '1', '1', '!', '1', ConsoleKeyEx.D1)); tell me the elements of this line. thanks. (i know my english is bad)

A good point of reference is the XMLDoc documentation [here](https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.System2/Keyboard/KeyMapping.cs#L73). 
https://github.com/CosmosOS/Cosmos/blob/92ea4ee64a6878910bd8e4de166f9882724eaca1/source/Cosmos.System2/Keyboard/KeyMapping.cs#L73-L88

**Comment by slavekrouta** [comment]
Im so sorry but I still dont understand. All I see is a chaos. Why are there so instances of the keymapping? Just explain. Thank you so much. I know I look kinda like an idiot, but i am a lil bit of idiot.

> > > Currently, the type of layout you're looking for isn't available in Cosmos. You can try implementing one yourself - if you do decide to do that, consider making a pull request that adds this functionality. We would love for you to join us and contribute to the project! You can see an example of such a layout [here](https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.System2/Keyboard/ScanMaps/US_Standard.cs#L9) - this class definition is for the standard US one.
> > 
> > 
> > I'll take a look, but can you tell me, what are these? : _keys.Add(new KeyMapping(0x02, '1', '!', '1', '1', '!', '1', ConsoleKeyEx.D1)); tell me the elements of this line. thanks. (i know my english is bad)
> 
> A good point of reference is the XMLDoc documentation [here](https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.System2/Keyboard/KeyMapping.cs#L73).
> 
> https://github.com/CosmosOS/Cosmos/blob/92ea4ee64a6878910bd8e4de166f9882724eaca1/source/Cosmos.System2/Keyboard/KeyMapping.cs#L73-L88
> 
> 
> Also note the constructor overloads; different key groups use different overloads in order to tidy up the code.

**Comment by ascpixi** [comment]
Don't worry, it's alright to ask for help! What are you specifically confused about? There are several overloaded constructors, and some of these constructors automatically fill in some of the values to tidy up the code (usually by repeating one of the values if the character doesn't change when some modifiers are active). The KeyMapping constructors take several different Unicode characters, each of which represents what character is produced when a key of a given scan code is pressed with the given modifiers.

Cosmos's scan codes mirror those of BIOS. You can find an exhaustive list of BIOS scan codes [here](https://www.fountainware.com/EXPL/bios_key_codes.htm).

So, for the overload I linked as an example, `aScanCode` represents the scan code of the **physical key on the keyboard.** `norm` represents what the Unicode character is without any modifiers, `shift` represents it when Shift is pressed, etc.

**Comment by slavekrouta** [comment]
@ascpixi I already started to make the cz layout. I got it like this: _keys.Add(new KeyMapping(0x09 - the key "code" num. 2, 'normal press', 'press+shift', 'press+altgr', 'press+caps', 'press+shift+caps', 'press+shift+numlock', ConsoleKeyEx.D8 - The key "code" num. 2));
am i right? Im sorry i know c# (well i guess) but this is the first time im trying to think 'bout os or keymapping :(

**Comment by ascpixi** [comment]
> @ascpixi I already started to make the cz layout. I got it like this: _keys.Add(new KeyMapping(0x09 - the key "code" num. 2, 'normal press', 'press+shift', 'press+altgr', 'press+caps', 'press+shift+caps', 'press+shift+numlock', ConsoleKeyEx.D8 - The key "code" num. 2)); am i right? Im sorry i know c# (well i guess) but this is the first time im trying to think 'bout os or keymapping :( But ill continue this way (when waiting for anser), for the good feeling.

Yep, that seems correct! You can always check out the existing keyboard layouts for guidance, and take a look at [this](https://github.com/CosmosOS/Cosmos/blob/master/source/Cosmos.System2/Keyboard/KeyboardManager.cs#L48) file to get a gist of what Cosmos will do with the values you provide.

**Comment by valentinbreiz** [comment]
Cosmos Gen3 has replaced the IL2CPU-based Gen2 codebase, and `gen3` is now the default branch of this repository. This issue was filed against Gen2, which is no longer developed, so it is being closed as part of the Gen3 cutover (#3088).

If you still hit this on Gen3, please open a fresh issue. The compiler, the build pipeline and the debugging story are all different now, so a new report will get a far more useful answer than reviving this one. Start here: https://cosmosos.github.io/

Gen2 remains on the `gen2` branch, and this issue stays readable.
